src/sys/windows/named_pipe.rs

```diff
@@ -874,9 +899,23 @@ fn read_done(status: &OVERLAPPED_ENTRY, events: Option<&mut Vec<Event>>) {
         match me.result(status.overlapped()) {
             Ok(n) => {
                 debug_assert_eq!(status.bytes_transferred() as usize, n);
-                buf.set_len(status.bytes_transferred() as usize);
+                // Extend the len depending on the initial len is necessary
+                // when we call `ReadFile` again after resizing
+                // our internal buffer
+                buf.set_len(buf.len() + status.bytes_transferred() as usize);
```

Comment:
This changes the call to append, which means we must also clear the buffer somewhere. I don't see a call to clear it.

Reply:
Do you mean as in resizing back the capacity to the default?

Reply:
Let's say I do two normal reads one after the other (of length N and M). I would expect the current code to do this:

1. We call `read_done` for the first read, changing the vector's length from 0 to N.
2. The N bytes are delivered to the user.
3. We call `read_done` again for the second read, changing the vector's length from N to N+M.
4. The user expects to get the M bytes, but instead gets the first N bytes again.

So I'm not thinking about the capacity here.

Reply:
What the current code does is:

1. `read_done` is invoked, changing the length from 0 to N
2. We determine that it's not the whole message, we call read again with a new state `InsufficientBufSize`
3. In this state, we will immediately call read again, but with a slightly different flow.
4. call `read_done` after the second read, length is now N + M.
5. The internal buffer holds N+M, the user can read it in any bytes they want. It doesn't have to be N or N+M.

Reply:
To clarify, I'm concerned about the situation where we *don't* get `InsufficientBufSize`. My example was intended to have two entirely separate reads.

Reply:
Not sure I entirely follow. We don't really change the situation when it's not `InsufficientBufSize`. the length of the buffer will always be 0 after we take it from the buffer pool, i.e. the append will only happen when it's `InsufficientBufSize`.

Maybe you meant that you were looking for this?

https://github.com/tokio-rs/mio/blob/af539876224e718233be0b141616b41787ff7b1a/src/sys/windows/named_pipe.rs#L1086-L1093

Reply:
Ah, yes, that's what I was missing. Thanks.